src/Polysemy/Internal/Final.hs

```diff
@@ -0,0 +1,191 @@
+{-# LANGUAGE TemplateHaskell #-}
+module Polysemy.Internal.Final where
+
+import Polysemy
+import Polysemy.Internal
+
+-----------------------------------------------------------------------------
+-- | This represents a function which produces
+-- an action of the final monad @m@ given:
+--
+--   * The initial effectful state at the moment the action
+--     is to be executed.
+--
+--   * A way to convert @z@ (which is typically @'Sem' r@)to @m@ by
+--     threading the effectful state through.
+--
+--   * An inspector that is able to view some value within the
+--     effectful state if the effectful state contains any values.
+--
+-- A @'Polysemy.Internal.Union.Weaving'@ provides these components,
+-- hence the name 'ThroughWeavingToFinal'.
+type ThroughWeavingToFinal m z a =
+     forall f
+   . Functor f
+  => f ()
+  -> (forall x. f (z x) -> m (f x))
+  -> (forall x. f x -> Maybe x)
+  -> m (f a)
+
+-----------------------------------------------------------------------------
+-- | An effect for embedding higher-order actions in the final target monad
+-- of the effect stack.
+--
+-- This is very useful for writing interpreters that interpret higher-order
+-- effects in terms of the final monad.
+--
+-- 'Final' is slightly weaker, but using it is otherwise more preferable
+-- than using a provided lowering function (a la 'Polysemy.Async.lowerAsync'):
+--
+--   * Interpreters using 'Final' may be run and composed like any other
+--     interpreter, unlike @lower-@ interpreters which need to be
+--     composed using '.@' or '.@@'.
+--
+--   * Multiple return-value changing interpreters using 'Final' may be
+--     composed together without issue. It's very difficult to
+--     compose 'Polysemy.Error.lowerError' with itself, unlike
+--     'Polysemy.Error.errorToIOFinal'.
+--
+--   * Initialization work of @lower-@ interpreters may be duplicated
+--     when composed together with '.@'. @-'Final'@ interpreters avoid
+--     this issue altogether, as long as the initialization work is performed
+--     outside of a 'interpretH' / 'Polysemy.Final.interpretFinal'.
+--
+--   * Instead of having access to a natural transformation
+--     @forall x. 'Sem' r x -> m x@, @'Final' m@ provides a
+--     distribution through the effectful state:
+--     @forall x. f ('Sem' r x) -> m (f x)@,
+--     together with the initial state @f ()@, and an @'Inspector' f@.
+--     (See 'ThroughWeavingToFinal')
+--     This is powerful enough for most purposes.
```

Comment:
Should this gigantic pros vs. cons compared to `lower-` interpreters stay? It made sense in `polysemy-zoo`, but does it in `polysemy` proper?

Reply:
Ok